contracts/src/lib/SupplySourceHelper.sol

```diff
@@ -54,31 +54,102 @@ library SupplySourceHelper {
+        uint256 value
+    ) internal returns (bool success, bytes memory ret) {
         if (supplySource.kind == SupplyKind.Native) {
-            Address.sendValue(payable(recipient), value);
+            success = sendValue(payable(recipient), value);
+            return (success, EMPTY_BYTES);
         } else if (supplySource.kind == SupplyKind.ERC20) {
-            IERC20(supplySource.tokenAddress).safeTransfer({to: recipient, value: value});
+            return ierc20Transfer(supplySource, recipient, value);
         }
     }
 
+    /// @notice Wrapper for an IERC20 transfer that bubbles up the success or failure
+    /// and the return value instead of reverting so a cross-message receipt can be
+    /// triggered from the execution.
+    /// This function the `safeTransfer` function used before.
+    function ierc20Transfer(
+        SupplySource memory supplySource,
+        address payable recipient,
+        uint256 value
+    ) internal returns (bool success, bytes memory ret) {
+        return
+            supplySource.tokenAddress.call(
+                // using IERC20 transfer instead of safe transfer so we can
+                // bubble-up the failure instead of reverting on failure so we
+                // can send the receipt.
+                abi.encodePacked(IERC20.transfer.selector, abi.encode(recipient, value))
+            );
+    }
+
     /// @notice Calls the target with the specified data, ensuring it receives the specified value.
-    function performCall(SupplySource memory supplySource, address payable target, bytes memory data, uint256 value) internal returns (bytes memory ret) {
+    function performCall(
+        SupplySource memory supplySource,
+        address payable target,
+        bytes memory data,
+        uint256 value
+    ) internal returns (bool success, bytes memory ret) {
         // If value is zero, we can just go ahead and call the function.
         if (value == 0) {
-            ret = Address.functionCall(target, data);
+            return functionCallWithValue(target, data, 0);
         }
 
         // Otherwise, we need to do something different.
         if (supplySource.kind == SupplyKind.Native) {
             // Use the optimized path to send value along with the call.
-            ret = Address.functionCallWithValue({target: target, data: data, value: value});
+            (success, ret) = functionCallWithValue({target: target, data: data, value: value});
         } else if (supplySource.kind == SupplyKind.ERC20) {
             // Transfer the tokens first, _then_ perform the call.
-            IERC20(supplySource.tokenAddress).safeTransfer({to: target, value: value});
-            ret = Address.functionCall(target, data);
+            (success, ret) = ierc20Transfer(supplySource, target, value);
+            if (success) {
+                // Perform the call only if the ERC20 was successful.
+                (success, ret) = functionCallWithValue(target, data, 0);
```

Comment:
Issue: the transfer + function call is not atomic. We need to figure out if putting these behind a function that uses revert semantics, and then calling it via delegatecall, would make these two actions atomic.